ext/re2/re2.cc

```diff
@@ -93,12 +94,85 @@ typedef struct {
   VALUE regexp, text;
 } re2_scanner;
 
-VALUE re2_mRE2, re2_cRegexp, re2_cMatchData, re2_cScanner;
+typedef struct {
+  RE2::Set *set;
+} re2_set;
+
+VALUE re2_mRE2, re2_cRegexp, re2_cMatchData, re2_cScanner, re2_cSet,
+      re2_eSetMatchError;
 
 /* Symbols used in RE2 options. */
 static ID id_utf8, id_posix_syntax, id_longest_match, id_log_errors,
           id_max_mem, id_literal, id_never_nl, id_case_sensitive,
-          id_perl_classes, id_word_boundary, id_one_line;
+          id_perl_classes, id_word_boundary, id_one_line,
+          id_unanchored, id_anchor_start, id_anchor_both;
+
+RE2::Options parse_re2_options(VALUE options) {
+  if (TYPE(options) != T_HASH) {
+    rb_raise(rb_eArgError, "options should be a hash");
+  }
+  VALUE utf8, posix_syntax, longest_match, log_errors,
+        max_mem, literal, never_nl, case_sensitive, perl_classes,
+        word_boundary, one_line;
+  RE2::Options re2_options;
+
+  utf8 = rb_hash_aref(options, ID2SYM(id_utf8));
+  if (!NIL_P(utf8)) {
+    re2_options.set_encoding(RTEST(utf8) ? RE2::Options::EncodingUTF8 : RE2::Options::EncodingLatin1);
+  }
+
+  posix_syntax = rb_hash_aref(options, ID2SYM(id_posix_syntax));
+  if (!NIL_P(posix_syntax)) {
+    re2_options.set_posix_syntax(RTEST(posix_syntax));
+  }
+
+  longest_match = rb_hash_aref(options, ID2SYM(id_longest_match));
+  if (!NIL_P(longest_match)) {
+    re2_options.set_longest_match(RTEST(longest_match));
+  }
+
+  log_errors = rb_hash_aref(options, ID2SYM(id_log_errors));
+  if (!NIL_P(log_errors)) {
+    re2_options.set_log_errors(RTEST(log_errors));
+  }
+
+  max_mem = rb_hash_aref(options, ID2SYM(id_max_mem));
+  if (!NIL_P(max_mem)) {
+    re2_options.set_max_mem(NUM2INT(max_mem));
+  }
+
+  literal = rb_hash_aref(options, ID2SYM(id_literal));
+  if (!NIL_P(literal)) {
+    re2_options.set_literal(RTEST(literal));
+  }
+
+  never_nl = rb_hash_aref(options, ID2SYM(id_never_nl));
+  if (!NIL_P(never_nl)) {
+    re2_options.set_never_nl(RTEST(never_nl));
+  }
+
+  case_sensitive = rb_hash_aref(options, ID2SYM(id_case_sensitive));
+  if (!NIL_P(case_sensitive)) {
+    re2_options.set_case_sensitive(RTEST(case_sensitive));
+  }
+
+  perl_classes = rb_hash_aref(options, ID2SYM(id_perl_classes));
+  if (!NIL_P(perl_classes)) {
+    re2_options.set_perl_classes(RTEST(perl_classes));
+  }
+
+  word_boundary = rb_hash_aref(options, ID2SYM(id_word_boundary));
+  if (!NIL_P(word_boundary)) {
+    re2_options.set_word_boundary(RTEST(word_boundary));
+  }
+
+  one_line = rb_hash_aref(options, ID2SYM(id_one_line));
+  if (!NIL_P(one_line)) {
+    re2_options.set_one_line(RTEST(one_line));
+  }
+
+  return re2_options;
```

Comment:
The extraction of this function seems to be causing an issue with versions of re2 with an ABI version of 0, see https://github.com/mudge/re2/runs/7521387298?check_suite_focus=true

```
/usr/include/re2/re2.h: In function ‘re2::RE2::Options parse_re2_options(VALUE)’:
/usr/include/re2/re2.h:675:5: error: ‘re2::RE2::Options::Options(const re2::RE2::Options&)’ is private
     Options(const Options&);
     ^
../../../../ext/re2/re2.cc:174:10: error: within this context
   return re2_options;
          ^
In file included from ../../../../ext/re2/re2.cc:10:0:
/usr/include/re2/re2.h: In function ‘VALUE re2_regexp_initialize(int, VALUE*, VALUE)’:
/usr/include/re2/re2.h:675:5: error: ‘re2::RE2::Options::Options(const re2::RE2::Options&)’ is private
     Options(const Options&);
     ^
../../../../ext/re2/re2.cc:751:57: error: within this context
     RE2::Options re2_options = parse_re2_options(options);
                                                         ^
In file included from ../../../../ext/re2/re2.cc:10:0:
/usr/include/re2/re2.h: In function ‘VALUE re2_set_initialize(int, VALUE*, VALUE)’:
/usr/include/re2/re2.h:676:10: error: ‘void re2::RE2::Options::operator=(const re2::RE2::Options&)’ is private
     void operator=(const Options&);
          ^
../../../../ext/re2/re2.cc:1437:17: error: within this context
     re2_options = parse_re2_options(options);
                 ^
In file included from ../../../../ext/re2/re2.cc:10:0:
/usr/include/re2/re2.h:676:10: error: ‘void re2::RE2::Options::operator=(const re2::RE2::Options&)’ is private
     void operator=(const Options&);
          ^
../../../../ext/re2/re2.cc:1439:17: error: within this context
     re2_options = RE2::DefaultOptions;
```

Reply:
Fixed in d6252548ab6ff39df35e56042df9d41ef99a44ac by passing `RE2::Options` by reference and mutating them in place rather than attempting to return a new instance.